File: app/src/main/java/com/jing/service/CoyoteActivity.java
package com.jing.service;

import android.annotation.SuppressLint;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothGatt;
import android.bluetooth.BluetoothGattCallback;
import android.bluetooth.BluetoothManager;
import android.bluetooth.le.BluetoothLeScanner;
import android.bluetooth.le.ScanCallback;
import android.bluetooth.le.ScanResult;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.widget.Button;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import com.jing.R;
import com.jing.toys.CoyoteTwo;
import com.jing.utils.BluetoothGattManager;
import com.jing.utils.BluetoothUtils;
import com.jing.utils.CoyoteConstant;
import com.jing.utils.ToastUtils;

/**
 * @Author：静
 * @Package：com.jing.service
 * @Project：灵静
 * @name：CoyoteActivity
 * @Date：2024/10/15 下午12:34
 * @Filename：CoyoteActivity
 * @Version：1.0.0
 */
public class CoyoteActivity extends AppCompatActivity {

    private BluetoothDevice mDevice;

    private String coyoteName;

    private String coyoteAddress;

    private BluetoothGatt mBluetoothGatt;

    private BluetoothUtils bluetoothUtils;

    private Button coyoteTwo, coyoteThree;


    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_coyote);

        coyoteTwo = findViewById(R.id.button2);
        coyoteThree = findViewById(R.id.button3);

        ActivityResultLauncher<Intent> intentActivityResultLauncher = registerForActivityResult(new ActivityResultContracts.StartActivityForResult(), result -> {
            bluetoothUtils.handleEnableBluetoothResult(result.getResultCode(), CoyoteActivity.this);
        });

        bluetoothUtils = new BluetoothUtils(this,intentActivityResultLauncher );

        bluetoothUtils.checkBluetoothAndRequestPermissions(this);

        coyoteTwo.setOnClickListener(v -> checkAndConnect("2.0"));
        coyoteThree.setOnClickListener(v -> checkAndConnect("3.0"));
    }

    @SuppressLint("MissingPermission")
    private void checkAndConnect(String version) {
        if (mBluetoothGatt != null && mBluetoothGatt.connect()) {
            if (mBluetoothGatt.getDevice().getName().equals(coyoteName) && mBluetoothGatt.getDevice().getAddress().equals(coyoteAddress) && version.equals("2.0")){
                Intent intent = new Intent(CoyoteActivity.this, CoyoteTwo.class);
                intent.putExtra("deviceName", coyoteName);
                intent.putExtra("deviceAddress", coyoteAddress);
                intent.putExtra("version", version);
                startActivity(intent);
            }

        } else {
            // 未连接，提醒用户
            ToastUtils.showToast(CoyoteActivity.this, "请先连接蓝牙设备");
        }
    }

    @Override
    protected void onResume() {
        super.onResume();
        startBluetoothScan();
    }

    @SuppressLint("MissingPermission")
    private void startBluetoothScan() {
        BluetoothManager bluetoothManager = (BluetoothManager) getSystemService(BLUETOOTH_SERVICE);
        BluetoothLeScanner bluetoothLeScanner = bluetoothManager.getAdapter().getBluetoothLeScanner();
        bluetoothLeScanner.startScan(new ScanCallback() {
            @Override
            public void onScanResult(int callbackType, ScanResult result) {
                super.onScanResult(callbackType, result);
                BluetoothDevice device = result.getDevice();
                if (device.getName() != null && device.getName().equals(CoyoteConstant.COYOTE_TWO_NAME)) {
                    mDevice = device;
                    coyoteName = device.getName();
                    coyoteAddress = device.getAddress();
                    ToastUtils.showToast(CoyoteActivity.this, "找到设备");
                    bluetoothLeScanner.stopScan(this);
                    connectToDevice(mDevice);
                }
            }

            @Override
            public void onScanFailed(int errorCode) {
                super.onScanFailed(errorCode);
                ToastUtils.showToast(CoyoteActivity.this, "未找到设备");
            }
        });
    }

    @SuppressLint("MissingPermission")
    private void connectToDevice(BluetoothDevice mDevice) {
       mBluetoothGatt= mDevice.connectGatt(this, false, new BluetoothGattCallback() {
            @Override
            public void onConnectionStateChange(BluetoothGatt gatt, int status, int newState) {
                if (newState == BluetoothGatt.STATE_CONNECTED) {
                    runOnUiThread(() -> ToastUtils.showToast(CoyoteActivity.this, "连接成功"));
                    BluetoothGattManager.getInstance().setBluetoothGatt(gatt);
                    gatt.discoverServices();
                } else if (newState == BluetoothGatt.STATE_DISCONNECTED) {
                    ToastUtils.showToast(CoyoteActivity.this, "连接断开");
                }
            }

           @Override
           public void onServicesDiscovered(BluetoothGatt gatt, int status) {
               if (status == BluetoothGatt.GATT_SUCCESS) {
                   displayAvailableServices(gatt);
               } else {
                   Log.w("CoyoteActivity", "服务发现失败: $status");
               }
           }
       });




    }

    private void displayAvailableServices(BluetoothGatt gatt) {
        Log.d("CoyoteActivity", "开始显示可用服务");
       gatt.getServices().forEach(service -> {
           Log.d("CoyoteActivity", "服务: " + service.getUuid().toString());
           service.getCharacteristics().forEach(characteristic -> {
               Log.d("CoyoteActivity", "特征: " + characteristic.getUuid().toString());
           });
       });
    }


    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        bluetoothUtils.handleEnableBluetoothResult(resultCode, this);
    }


}

File: app/src/main/java/com/jing/toys/CoyoteActivity.kt
package com.jing.toys

import android.annotation.SuppressLint
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothGattCallback
import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothManager
import android.bluetooth.BluetoothProfile
import android.content.Context
import android.os.Bundle
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.jing.R
import com.jing.utils.BluetoothUtils
import java.util.UUID

/**
 *@Author：静
 *@Package：com.jing
 *@Project：灵境
 *@Date：2024/10/5  下午11:59
 *@Filename：HomeActivity
 *@Description：郊狼
 *@Version：1.0.0
 */
class CoyoteActivity : AppCompatActivity() {

    private lateinit var bluetoothGatt: BluetoothGatt
    private lateinit var batteryLevelTextView: TextView
    @SuppressLint("MissingPermission")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_coyote)
        batteryLevelTextView = findViewById(R.id.batteryLevelText)
        BluetoothUtils.init(this)
        BluetoothUtils.checkAndRequestPermissions(this)

        val bluetoothManager = this.getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager
        val bluetoothAdapter = bluetoothManager.adapter
        val pairedDevices: Set<BluetoothDevice> = bluetoothAdapter.bondedDevices

        if (pairedDevices.isNotEmpty()) {
            for (device in pairedDevices) {
                val deviceName = device.name
                val deviceAddress = device.address
                bluetoothGatt = device.connectGatt(this, false, gattCallback)
                break // 如果只需连接第一个设备，可以在这里跳出循环
            }
        } else {
            Toast.makeText(this, "未配对设备", Toast.LENGTH_SHORT).show()
        }
    }

    private val gattCallback = object : BluetoothGattCallback() {
        @SuppressLint("MissingPermission")
        override fun onConnectionStateChange(gatt: BluetoothGatt?, status: Int, newState: Int) {
            if (newState == BluetoothProfile.STATE_CONNECTED) {
                // 连接成功后发现服务
                gatt?.discoverServices()
            }
        }

        @SuppressLint("MissingPermission")
        override fun onServicesDiscovered(gatt: BluetoothGatt?, status: Int) {
            if (status == BluetoothGatt.GATT_SUCCESS) {
                // 获取电量特征
                val batteryLevelCharacteristic: BluetoothGattCharacteristic? =
                    gatt?.getService(UUID.fromString("955A180A-0FE2-F5AA-A094-84B8D4F3E8AD"))
                        ?.getCharacteristic(UUID.fromString("955A1500-0FE2-F5AA-A094-84B8D4F3E8AD"))

                batteryLevelCharacteristic?.let {
                    // 读取电量
                    gatt.readCharacteristic(it)
                }
            }
        }

        override fun onCharacteristicRead(gatt: BluetoothGatt?, characteristic: BluetoothGattCharacteristic?, status: Int) {
            if (status == BluetoothGatt.GATT_SUCCESS) {
                val batteryLevel: Int = characteristic?.getIntValue(BluetoothGattCharacteristic.FORMAT_UINT8, 0) ?: 0
                runOnUiThread {
                    batteryLevelTextView.text = "电量: $batteryLevel%"
                }
            }
        }
    }
}
